[FSDP] Optim state chkpt: key by param name, not ID

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74912 [FSDP] Add re-key btw param names/IDs for optim state dict
* **#74879 [FSDP] Optim state chkpt: key by param name, not ID**
* #74215 [FSDP] Add full optim state dict
* #74834 [Easy][FSDP] (Reland) Doc fixes

**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

TTS contribution: ~80 seconds (timed on AWS cluster).

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

Commit:
[FSDP] Optim state chkpt: key by param name, not ID

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).

zhaojuanmao commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -455,42 +459,110 @@ def _flatten_zero_dim_tensor_optim_state(
+    return flat_param_to_fsdp_module
+
+
+def _get_param_to_unflat_param_names(
```
looks like this is a useful generic utils that can be used in other places, let's move to utils.py?

awgu replied:
There is some challenge with circular imports for moving to `utils.py` since this function uses `FullyShardedDataParallel` and `FSDP_WRAPPED_MODULE` from `fully_sharded_data_parallel.py` and `FlatParameter` and `FPW_MODULE` from `flatten_params_wrapper.py`, while both of those files import from `utils.py`.
Maybe we keep `utils.py` for utilities agnostic to `FSDP`, and I will move `_get_param_to_unflat_param_names()` to the bottom of `fully_sharded_data_parallel.py`.

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

TTS contribution: ~80 seconds (timed on AWS cluster).

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).

Commit:
Update on "[FSDP] Optim state chkpt: key by param name, not ID"


**Overview**
This replaces keying by parameter ID with keying by full parameter name. So far, this seems to allow users to change the `FSDP` wrapping scheme between save and load time, which was unsupported when keying by parameter ID.

In order to key by full parameter name, we must construct those names, which involves cleaning any `FSDP`-specific prefixes. To do this, I had to introduce some overlap with the model state dict logic since that also performs this functionality. It is difficult to refactor the commonality because the overlap is embedded in `torch.nn.Module.state_dict()` (in how it prepends module names as prefixes as it recurses into nested submodules).

**Test Plan**
The tests that were meant to raise an error when changing the `FSDP` wrapping scheme are no longer raising errors. 

TTS contribution: ~80 seconds (timed on AWS cluster).

Differential Revision: [D35203958](https://our.internmc.facebook.com/intern/diff/D35203958)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203958).
